controller/volume_controller.go

```diff
@@ -2688,11 +2688,59 @@ func (c *VolumeController) upgradeEngineForVolume(v *longhorn.Volume, es map[str
 		return nil
 	}
 
-	// only start live upgrade if volume is healthy
 	if v.Status.State != longhorn.VolumeStateAttached || v.Status.Robustness != longhorn.VolumeRobustnessHealthy {
+		if v.Status.State != longhorn.VolumeStateAttached || v.Status.Robustness != longhorn.VolumeRobustnessDegraded {
+			return nil
+		}
+		// Clean up inactive replica corresponding to non-existing active replica.
+		// This will allow volume controller to schedule and rebuild a new active replica
+		// and make the volume healthy again for the live upgrade to be able to continue.
+		// https://github.com/longhorn/longhorn/issues/7012
+		hasCorrespondingActiveReplica := func(replica *longhorn.Replica) bool {
+			for _, r := range rs {
+				if r.Spec.Active &&
+					r.Name != replica.Name &&
+					r.Spec.NodeID == replica.Spec.NodeID &&
+					r.Spec.DiskID == replica.Spec.DiskID &&
+					r.Spec.DiskPath == replica.Spec.DiskPath &&
+					r.Spec.DataDirectoryName == replica.Spec.DataDirectoryName {
+					return true
+				}
+			}
+			return false
+		}
+
+		// sort the replica by name so that we don't select multiple replicas when running this logic repeatedly quickly
```

Comment:
Can you clarify this? I wasn't able to fully understand why the sorting was necessary.

Reply:
I think it is the same reason as the other uses of sort in this controller - if we run this code twice with the same set of replicas, we want it to be idempotent, so we better choose the same inactive replica to delete.

Reply:
Yeah, if there are multiple inactive replicas to be deleted, I would want to delete the same one, one by one, at a time. Effectively slowing down the replica deletion.